src/Language/PureScript/Docs/Convert.hs

```diff
@@ -55,8 +63,79 @@ convertModule externs env checkEnv =
 insertValueTypesAndAdjustKinds ::
   P.Environment -> Module -> Module
 insertValueTypesAndAdjustKinds env m =
-  m { modDeclarations = map go (modDeclarations m) }
+  m { modDeclarations = map (go . insertInferredRoles . convertFFIDecl) (modDeclarations m) }
   where
+  -- |
+  -- Convert FFI declarations into data declaration
+  -- by generating the type parameters' names based on its kind signature.
+  -- Note: `Prim` modules' docs don't go through this conversion process
+  -- so `ExternDataDeclaration` values will still exist beyond this point.
+  convertFFIDecl d@Declaration { declInfo = ExternDataDeclaration kind roles } =
+    d { declInfo = DataDeclaration P.Data (genTypeParams kind) roles
+      , declKind = Just (KindInfo P.DataSig kind)
+      }
+
+  convertFFIDecl other = other
+
+  insertInferredRoles d@Declaration { declInfo = DataDeclaration dataDeclType args [] } =
+    d { declInfo = DataDeclaration dataDeclType args inferredRoles }
+
+    where
+    inferredRoles :: [P.Role]
+    inferredRoles = do
+      let key = P.Qualified (Just (modName m)) (P.ProperName (declTitle d))
+      case Map.lookup key (P.types env) of
+        Just (_, tyKind) -> case tyKind of
+          P.DataType _ tySourceTyRole _ ->
+            map (\(_,_,r) -> r) tySourceTyRole
+          P.ExternData rs ->
+            rs
+          _ ->
+            []
+        Nothing ->
+          err $ "type not found: " <> show key
+
+  insertInferredRoles other =
+    other
+
+  -- |
+  -- Given an FFI declaration like this
+  -- ```
+  -- foreign import data Foo
+  --    :: forall a b c d
+  --     . MyKind a b
+  --    -> OtherKind c d
+  --    -> Symbol
+  --    -> (Type -> Type)
+  --    -> (Type) -- unneeded parens a developer typo
+  --    -> Type
+  -- ```
+  -- Return a list of values, one for each implicit type parameter
+  -- of `(tX, Nothing)` where `X` refers to the index of he parameter
+  -- in that list, matching the values expected by `Render.toTypeVar`
+  genTypeParams :: Type' -> [(Text, Maybe Type')]
+  genTypeParams kind = do
+    let n = countParams 0 kind
+    if n == 0
+      then []
+      else map (\i -> ("t" <> T.pack (show i), Nothing)) [0..n]
+    where
+      countParams :: Integer -> Type' -> Integer
+      countParams acc = \case
+        P.ForAll _ _ _ rest _ ->
+          countParams acc rest
+
+        P.TypeApp _ f a@(P.TypeApp _ _ _) | isFunctionApplication f ->
```

Comment:
We should not match on type applications here, otherwise we count 0 instead of 1 for `Type -> Type`. The range syntax being inclusive `genTypeParams` would then yield lists of length n+1 for kinds of arity n though! I think we should take from an infinite list instead, which would also simplify away the conditional expression:

```suggestion
   (\(i :: Int) -> ("t" <> T.pack (show i), Nothing)) <$> take n [0..]
    where
      countParams :: Int -> Type' -> Int
      countParams acc = \case
        P.ForAll _ _ _ rest _ ->
          countParams acc rest

        P.TypeApp _ f a | isFunctionApplication f ->
```

Reply:
I ran out of time today, so I'll have to try this out later. I think my code above does work for the case you mentioned, but I think your code is cleaner overall. Thanks for the feedback btw on this PR and the others!